Implement kj::newTee()

harrishancock opened the pull request:
`kj::newTee()` accepts an `Own<AsyncInputStream>` and returns two `Own<AsyncInputStream>`s which provide access to the same data as the original.

The two new `Own<AsyncInputStream>` objects are called branches of the tee. Each branch may have a sink. A caller may create a sink on a branch by calling `tryRead()` or `pumpTo()` on it. The returned promise *is* the sink.

Sinks created by `tryRead()` provide vacuum, whereas sinks created by `pumpTo()` do not. If a tee has at least one branch with a sink with vacuum, it will wait for any backpressure from any `pumpTo()` sinks to subside, then perform an inner read to try to fill the vacuum. This means that in a pristine state, the first `tryRead()` finishes as soon as data is available (because the pump has zero backpressure), but subsequent `tryRead()`s wait for the backpressure of other branches' `pumpTo()`s caused by the initial inner read.

While `pumpTo()` sinks do not provide vacuum, if there is ever a point at which all branches have active sinks (of any type), the tee considers there to be vacuum, and pumps freely, respecting the backpressure of the slowest sink.

If a branch has no active sink when the inner read finishes (either because one was never created, or a promise was canceled), the newly-read data is buffered for that branch, instead.

An exception encountered by the inner read prevents further inner reads. The encountered exception is propagated to all branches once their sinks have reached the end of their buffer. I decided to err on the side of filling `tryRead()` sinks as much as possible rather than eagerly rejecting them: if a read exception occurs, `tryRead()` sinks will see short reads *before* the exception. Similarly, `pumpTo()` sinks will pump all buffered data before seeing the exception. However, `pumpTo()` promises are rejected immediately upon flushing their buffer (and waiting for backpressure).

EOF propagation follows the same rules as exception propagation, except that instead of rejecting promises, it fulfills them with zero.

An exception encountered during a `write()` operation inside of `pumpTo()` rejects that branch's `pumpTo()` promise, but does not propagate to the other branch. Canceling a `pumpTo()` promise cancels any `write()` promises it has in flight.

harrishancock commented:
CI passes!

harrishancock commented on c++/src/kj/async-io.c++:
```diff
@@ -1125,6 +1126,633 @@ TwoWayPipe newTwoWayPipe() {
+#if __cplusplus < 201402L
+      inline constexpr Need(uint64_t minBytes, uint64_t maxBytes)
+          : minBytes(minBytes), maxBytes(maxBytes) {}
+      // TODO(cleanup): This constructor is only here to support brace initialization in C++11. It
```
gcc 4.9 doesn't seem to implement this part of C++14. :( I'll adjust the comment here (and where I copied it from).

harrishancock replied:
gcc 4.9.4 sets `__cplusplus` to `201300L` even with `-std=c++14`: https://gcc.godbolt.org/z/7eidrX

harrishancock commented on c++/src/kj/async-io.c++:
```diff
@@ -1125,6 +1126,633 @@ TwoWayPipe newTwoWayPipe() {
+
+class AsyncTee final: public Refcounted {
+public:
+  using BranchId = uint8_t;
```
Yup, in this case it doesn't really matter, will use `uint`.

harrishancock commented on c++/src/kj/async-io.h:
```diff
@@ -203,6 +203,21 @@ struct CapabilityPipe {
+struct Tee {
+  // Two AsyncInputStreams which each read the same data from some wrapped inner AsyncInputStream.
+
+  Own<AsyncInputStream> ins[2];
```
I like `branches`. My initial stab at this returned one `out` and two `ins` -- modeled more after a `newOneWayPipe()`'s return struct -- so the `ins` is a bit vestigial.

kentonv commented on c++/src/kj/async-io.c++:
```diff
@@ -1261,8 +1264,8 @@ private:
           : minBytes(minBytes), maxBytes(maxBytes) {}
-      // TODO(cleanup): This constructor is only here to support brace initialization in C++11. It
-      //   should be removed once we upgrade to C++14.
+      // TODO(cleanup): This constructor is only here to support brace initialization in gcc 4.9. It
```
:man_facepalming: If you feel like bumping the CI to a newer GCC now I'm cool with that. Or if you'd rather leave it, that's fine too.

harrishancock commented:
@kentonv I think this is good to go. Squashed fixups and updated Travis to use gcc 5.0.

I noticed that the installation page on capnproto.org still mentions VS2015, gcc 4.x, and C++11, btw. I refrained from updating those in this PR since I recall you mentioning that updating the web page is sort of a pain.
